tests/Feature/CountryControllerTest.php

```diff
@@ -28,26 +28,75 @@ public function testCountryStore(): void
     {
         $data = Country::factory()->make()->toArray();
 
-        $this->post(route("countries.store"), $data);
+        $this->post(route(name:"countries.store"), data: $data);
 
-        $this->assertDatabaseHas("countries", $data);
+        $this->assertDatabaseHas(table:"countries", data: $data);
+    }
+
+    public function testCountryStoreIsRefusingDuplicate(): void
+    {
+        $country = Country::factory()->make()->toArray();
+
+        Country::query()->create([
+            "name" => $country["name"],
+            "latitude" => "-55.54323",
+            "longitude" => "42.3721",
+            "iso" => $country["iso"],
+        ])->toArray();
+
+        $this->post(route(name:"countries.store"), data: $country);
+
+        $this->assertDatabaseMissing(table: "countries", data: $country);
+    }
+
+    public function testCountryStoreIsRefusingInvalidData(): void
+    {
+        $invalidCountry = Country::factory()->make([
+            "latitude" => "string",
+        ])->toArray();
+
+        $this->post(route(name: "countries.store"), data: $invalidCountry);
+        $this->assertDatabaseMissing(table: "countries", data: $invalidCountry);
     }
 
     public function testCountryUpdate(): void
     {
         $data = Country::factory()->make()->toArray();
 
-        $this->post(route("countries.store"), $data);
+        $this->post(route(name: "countries.store"), data: $data);
 
-        $this->assertDatabaseHas("countries", $data);
+        $this->assertDatabaseHas(table:"countries", data: $data);
+    }
+
+    public function testCountryUpdateIsRefusingInvalidData(): void
+    {
+        $this->testCountryUpdateProperties(name: "poland");
+        $this->testCountryUpdateProperties(longitude: "string");
+        $this->testCountryUpdateProperties(latitude: "string");
+        $this->testCountryUpdateProperties(iso: "Pl");
```

Comment:
Use PHPUnit dataprovider for that. (the best option tip number 7, with Generator)

https://blog.martinhujer.cz/how-to-use-data-providers-in-phpunit/

Reply:
DataProviders are not trick, it is used for such cases.

https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers
https://docs.phpunit.de/en/10.2/annotations.html#dataprovider